torch/csrc/monitor/counters.h

```diff
@@ -104,88 +103,138 @@ class Stat {
     }
 
     current_.count += 1;
-    if (windowSize_ > 0 && current_.count >= windowSize_) {
-      saveCurrentLocked();
-    }
+    maybeLogLocked();
   }
 
   const std::string& name() const noexcept {
     return name_;
   }
 
-  int64_t windowSize() const noexcept {
-    return windowSize_;
-  }
-
   // count returns the number of items in the current open window.
   int64_t count() noexcept {
     std::lock_guard<std::mutex> guard(mu_);
 
     return current_.count;
   }
 
-  // closeWindow finalizes the collected stats window so they can be accessed
-  // via get().
-  // If the Stat has a windowSize specified this doesn't do anything since the
-  // window is automatically closed when enough samples have been logged.
-  void closeWindow() noexcept {
-    if (windowSize_ <= 0) {
-      std::lock_guard<std::mutex> guard(mu_);
+  std::unordered_map<Aggregation, T> get() noexcept {
+    std::lock_guard<std::mutex> guard(mu_);
+    return getLocked();
+  }
+
+ protected:
+  virtual void maybeLogLocked() = 0;
+
+  void logLocked() {
+    prev_ = current_;
+    current_ = Values();
+
+    // don't log event if there's no data
+    if (prev_.count == 0) {
+      return;
+    }
 
-      saveCurrentLocked();
+    Event e;
+    e.type = "torch.monitor.Stat";
+    e.message = name_;
+    e.timestamp = std::chrono::system_clock::now();
+
+    auto stats = getLocked();
+    e.metadata.reserve(stats.size());
+    for (auto& kv : stats) {
+      std::stringstream key;
+      key << name_;
+      key << ".";
+      key << aggregationName(kv.first);
+      e.metadata[key.str()] = kv.second;
     }
+
+    logEvent(e);
```

Comment:
having counters get published like Events is a fair choice for structured logging services (e.g. Scuba), but for simple timeseries based metric logging services (e.g. ods), the OdsEventHandler would have to just drop a bunch of fields (e.g. type) and just read off `event.metadata["foo.bar.AVG"] -> value`. But how does it know which key in the metadata map to read? would it iterate through all of them and only publish the ones where the value is a "number" type?